### PR TITLE
Multiplication after division

### PR DESCRIPTION
**Pull request by gmtesla**
This PR updates math operations to do multiplication first before division during the calculations to avoid rounding issues of Solidity math.
```
e.g 25% apply 
_precision * 25 / 100 => _precision / 4
// It is still human-readable and it has a comment to notify it applies 25%

reward = providerPools[tokenAddress] / leverage * _rewardPercentage / _precision
=>
reward = providerPools[tokenAddress] * _rewardPercentage / leverage / _precision
```

### CODE REVIEW

**Review by wilhempujar — CHANGES_REQUESTED** [review]
OK but instead of using a hardcode 4, best to use a constant.

**Comment by gmtesla** [comment]
> OK but instead of using a hardcode 4, best to use a constant.
I don't think that's good.
If we define a constant value, when calculating the getDeviation, it will access to the storage to get this constant and it will take more gas fees.

**Review by bussyjd — APPROVED** [review]
LGTM

File: contracts/SLORegistry.sol [hunk 1]
```diff
@@ -17,18 +17,19 @@ contract SLORegistry {
...
-     * @param sla 1. -
-     * @param sloValue 2. -
-     * @param sloType 3. -
-     */
+
+    /// @dev SLO Registered event
     event SLORegistered(address indexed sla, uint256 sloValue, SLOType sloType);
 
+    /// @notice maximum cap of deviation percent = 25%, base 10000
+    uint16 private constant deviationCapPercent = 2500;
```
Comment (wilhempujar) [inline contracts/SLORegistry.sol]: `deviationCap` better?

We're not using the "percent" suffix for any of the percentage in the protocol.

File: contracts/Staking.sol [hunk 2]
```diff
@@ -335,11 +333,11 @@ contract Staking is Ownable, ReentrancyGuard {
...
             address tokenAddress = allowedTokens[index];
 
-            uint256 reward = ((providersPool[tokenAddress] / leverage) *
-                _rewardPercentage) / _precision;
+            uint256 reward = (providersPool[tokenAddress] * _rewardPercentage) /
+                (leverage * _precision);
 
             // Reward must be less than 25% of usersPool to ensure payout at all time
-            if (reward > (usersPool[tokenAddress] * 25) / 100) {
+            if (reward > usersPool[tokenAddress] / 4) {
```
Comment (wilhempujar) [inline contracts/Staking.sol]: Should use `deviationCap`

File: contracts/Staking.sol [hunk 3]
```diff
@@ -377,10 +375,9 @@ contract Staking is Ownable, ReentrancyGuard {
                 _rewardPercentage) / _precision;
 
             // Compensation must be less than 25% of providersPool to ensure payout at all time
-            if (compensation > (providersPool[tokenAddress] * 25) / 100) {
+            if (compensation > providersPool[tokenAddress] / 4) {
```
Comment (wilhempujar) [inline contracts/Staking.sol]: Should use `deviationCap`

File: contracts/Staking.sol [hunk 4]
```diff
@@ -81,6 +79,7 @@ contract Staking is Ownable, ReentrancyGuard {
     /// @dev claiming fees when a user claim tokens, base 10000
     uint16 private constant ownerRewardsRate = 30; // 0.3%, base 10000
     uint16 private constant protocolRewardsRate = 15; // 0.15%, base 10000
+    uint16 private constant rewardsCapRate = 2500;
```
Comment (wilhempujar) [inline contracts/Staking.sol]: Don't see this being used anywhere (yet)?

**Review by wilhempujar — APPROVED** [review]
LGTM

**Pull request merged**